lib/node_modules/@stdlib/_tools/remark/plugins/remark-lint-html-section-structure/lib/main.js

```diff
@@ -0,0 +1,366 @@
+	var structure;
+	var positions;
+	var className;
+	var errCode;
+	var msg;
+	var i;
+
+	debug( 'Linting file: %s', file.path || '' );
+
+	sectionStack = [];
+	rootSections = [];
+	structure = [];
+	positions = {};
+
+	visit( tree, 'html', visitor );
+
+	debug( 'Final section stack length: %d', sectionStack.length );
+
+	if ( sectionStack.length > 0 ) {
+		debug( 'Found unclosed sections: %d', sectionStack.length );
+		for ( i = 0; i < sectionStack.length; i++ ) {
+			className = sectionStack[ i ].className || '[no class specified]';
+			msg = 'Unclosed section tag. Missing </section> closing tag for section with class="' + className + '". Add a closing tag to fix.';
+			errCode = 'unclosed-section';
+			reportErr( file, sectionStack[ i ], msg, errCode );
+		}
+	}
+
+	debug( 'Finished linting: %s', file.path || '' );
+	return clbk();
+
+	/**
+	* Callback invoked upon finding a matching node.
+	*
+	* @private
+	* @param {Object} node - AST node
+	* @param {number} index - position of `node` in `parent`
+	* @param {Object} parent - parent AST node
+	* @returns {void}
+	*/
+	function visitor( node, index, parent ) {
+		var combinedMatch;
+		var commentClass;
+		var sectionMatch;
+		var commentNode;
+		var nextNode;
+		var endMatch;
+		var current;
+		var match;
+		var data;
+
+		// Store position for tracking purposes:
+		if ( node.position ) {
+			positions[ node.position.start.line ] = node;
+		}
+
+		// Check if this is a combined section end with comment on the same line (which is an error):
+		combinedMatch = SECTION_END_WITH_COMMENT.exec( node.value );
+		if ( combinedMatch ) {
+			debug( 'Found combined section end with comment: %s', node.value );
+			msg = 'Missing empty line between closing section tag and comment. There should be an empty line between </section> and <!-- /.' + combinedMatch[1] + ' -->';
+			errCode = 'missing-empty-line';
+			reportErr( file, node, msg, errCode );
+
+			// Make sure to pop the section stack for this case - otherwise we'll get "unclosed section" errors:
+			if ( sectionStack.length > 0 ) {
+				debug( 'Popping section stack for combined section end+comment' );
+				sectionStack.pop();
+			}
+			return;
+		}
+
+		// Check if this is a section start tag
+		sectionMatch = SECTION_START.exec( node.value );
+		if ( sectionMatch ) {
+			debug( 'Found a section start tag with class: %s', sectionMatch[1] || '[none]' );
+
+			data = {
+				'node': node,
+				'className': sectionMatch[1] || '',
+				'position': node.position,
+				'children': [],
+				'index': index,
+				'parent': parent
+			};
+
+			if ( sectionStack.length === 0 ) {
+				// This is a root level section...
+				rootSections.push( data );
+				structure.push( data );
+			} else {
+				// This is a nested section, add to parent's children:
+				sectionStack[ sectionStack.length - 1 ].children.push( data );
+			}
+
+			sectionStack.push( data );
+			return;
+		}
+
+		// Check if this is a section end tag:
+		endMatch = SECTION_END.exec( node.value );
```

Comment:
Since we are using `exec`, do we need to worry about resetting the regexp state?

Reply:
 No. [Only stateful ](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/RegExp/exec#description) when global or sticky flags are set.